streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -163,7 +172,8 @@ public KafkaStreams(TopologyBuilder builder, StreamsConfig config, KafkaClientSu
             this.threads[i] = new StreamThread(builder, config, clientSupplier, applicationId, clientId, processId, metrics, time);
             storeProviders.add(new StreamThreadStateStoreProvider(threads[i]));
         }
-
+        kafkaStreamsInstances = new StreamsMetadataState(builder);
+        threads[0].setPartitionsByHostStateChangeListener(kafkaStreamsInstances);
```

Comment:
If we are always setting this listener, could we just remove this class and make the `StreamsMetadataState` shared among all threads to be updated in the rebalance callback with access locks (in practice we would recommend on thread per instance so this should be OK).